### PR TITLE
Support constraint imperfect distributions (#6811)

### PR DESCRIPTION
**Pull request by YilouWang**
## Summary
Support variable (non-constant) weight expressions in `dist` constraints (IEEE 1800-2017 §18.5.4). Previously, dist constraints with variable weights were lowered to simple `inside`-style boolean expressions, discarding weight information. Now they are lowered to a weighted bucket selection via `AstConstraintIf` chains that preserve correct probability distributions at runtime.

## Issue
Fixes verilator/verilator#6811

## Changes
- `V3AstNodeExpr.h`: Fix `AstDist` constructor to use `ASTGEN_SUPER_Dist` instead of `ASTGEN_SUPER_Inside`
- `V3Width.cpp` (`visit(AstDist*)`): Replace blanket `CONSTRAINTIGN` warning with a lightweight guard that passes simple scalar/range dist to `V3Randomize` while still lowering container-type dist (queues, arrays) via `insideItem()`
- `V3Randomize.cpp` (`lowerDistConstraints`): New function that transforms `AstDist` into weighted bucket selection -- builds expression-based cumulative sums, a random bucket variable, and a backward `AstConstraintIf` chain. Supports both `:=` (per-value) and `:/` (divided) weight operators with constant or variable weights
- `V3Randomize.cpp`: Add `m_distNum` counter for unique temp variable naming across multiple dist constraints
- `t_randomize.py`: Update from lint-fail to pass test (removed CONSTRAINTIGN warning)
- `t_randomize.out`: Deleted (no longer needed)

## Test
- Added `test_regress/t/t_constraint_dist_weight.{v,py}` -- 5 scenarios: constant scalar weights (`:=`), constant range weights (`:/`), zero-weight exclusion, variable scalar weights, variable range weights. Each runs 2000 iterations with statistical distribution checks.

---
Developed by PlanV GmbH, assisted with Claude Code.

Reviewed by YilouWang.

### CODE REVIEW

File: src/V3Randomize.cpp [hunk 1]
```diff
@@ -3498,6 +3660,7 @@ class RandomizeVisitor final : public VNVisitor {
         VL_RESTORER(m_randCaseNum);
         m_modp = nodep;
         m_randCaseNum = 0;
+        m_distNum = 0;
```
Comment (wsnyder) [inline src/V3Randomize.cpp:3723]: Need a VL_RESTORER(m_distNum) just above.

File: test_regress/t/t_constraint_dist_weight.v [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+    // := scalar weights: expect ~75% for value 255
+    sc = new;
+    count_high = 0;
+    repeat (total) begin
+      `checkd(sc.randomize(), 1);
+      if (sc.x == 8'd255) count_high++;
+      else `checkd(sc.x, 0);
+    end
+    `check_range(count_high, 1200, 1800);
```
Comment (wsnyder) [inline test_regress/t/t_constraint_dist_weight.v]: Suggest make these percentages of count.

```suggestion
    `check_range(count_high, count * 0.6, count * 0.8);
```

File: test_regress/t/t_randomize.py [hunk 3]
```diff
@@ -11,6 +11,6 @@
 
 test.scenarios('vlt')
 
-test.lint(fails=True, expect_filename=test.golden_filename)
+test.lint()
```
Comment (wsnyder) [inline test_regress/t/t_randomize.py]: If this now compiles should be a test.compile + test.execute.

File: test_regress/t/t_constraint_dist_weight.v [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+class DistRange;
+  rand bit [7:0] x;
+  constraint c { x dist { [8'd0:8'd9] :/ 1, [8'd10:8'd19] :/ 3 }; }
+endclass
+
+class DistZeroWeight;
+  rand bit [7:0] x;
+  constraint c { x dist { 8'd0 := 0, 8'd1 := 1, 8'd2 := 1 }; }
+endclass
+
```
Comment (wsnyder) [inline test_regress/t/t_constraint_dist_weight.v:27]: Add a test with just a zero weight also please.

File: src/V3Randomize.cpp [hunk 5]
```diff
@@ -3480,6 +3479,169 @@ class RandomizeVisitor final : public VNVisitor {
...
+            }
+
+            // Store totalWeight in temp var (evaluated once, used twice)
+            const int distId = m_distNum++;
+            const std::string totalName = "__Vdist_total" + cvtToStr(distId);
+            AstVar* const totalVarp
+                = new AstVar{fl, VVarType::BLOCKTEMP, totalName, taskp->findUInt64DType()};
+            totalVarp->noSubst(true);
+            totalVarp->lifetime(VLifetime::AUTOMATIC_EXPLICIT);
+            totalVarp->funcLocal(true);
```
Comment (wsnyder) [inline src/V3Randomize.cpp:3615]:
```suggestion
            totalVarp->funcLocal(true);
            totalVarp->isInternal(true);
```

File: src/V3Randomize.cpp [hunk 6]
```diff
@@ -3480,6 +3479,169 @@ class RandomizeVisitor final : public VNVisitor {
...
+                } else {
+                    AstNodeExpr* const condp
+                        = new AstLte{fl, new AstVarRef{fl, bucketVarp, VAccess::READ}, cumSums[i]};
+                    chainp = new AstConstraintIf{fl, condp, thenp, chainp};
+                }
+            }
+
+            if (chainp) {
+                constrExprp->replaceWith(chainp);
+                VL_DO_DANGLING(constrExprp->deleteTree(), constrExprp);
```
Comment (wsnyder) [inline src/V3Randomize.cpp]: Use pushDeletep instead of deleteTree please, helps debug

**Comment by YilouWang** [comment]
@wsnyder Thanks for the review! All feedback has been addressed in the latest commits. Here's a summary of the additional changes beyond the original dist weight support:

1. Added **all-zero weight** test case (DistAllZeroWeight) -- when all weights are zero the dist constraint becomes unconstrained, randomize() still succeeds
2. To make `t_randomize `fully pass (compile + execute), two related issues surfaced and were fixed:

- `disable soft` support: Added VlRandomizer::disable_soft() which removes soft constraints by variable name
- Bare variable constraint fix: multi-bit expressions used as bare constraints now correctly wrap as expr != 0. This was exposed by the dist + soft combination causing a regression in constraint if/else handling
